src/src/com/microsoft/aad/adal/StorageHelper.java

```diff
@@ -305,70 +283,241 @@ public String decrypt(String value) throws GeneralSecurityException, IOException
+                // handle it as regular KeyStoreException
+                ClientAnalytics.logEvent(new AndroidKeyStoreFailureEvent(
+                        new InstrumentationPropertiesBuilder(e)));
+                throw new KeyStoreException(e);
+            }
+        }
+        
+        throw new KeyStoreException("KeyPair entry does not exist.");
+    }
+    
+    /**
+     * Check if KeyPair exists on AndroidKeyStore. 
+     */
+    private synchronized boolean doesKeyPairExist() throws GeneralSecurityException, IOException {
+        KeyStore keyStore = KeyStore.getInstance("AndroidKeyStore");
+        keyStore.load(null);
+        
+        final boolean isKeyStoreCertAliasExisted;
+        try {
+            isKeyStoreCertAliasExisted = keyStore.containsAlias(KEY_STORE_CERT_ALIAS);
+        } catch (final NullPointerException exception) {
+            // There is an issue with Android Keystore when remote service attempts 
+            // to access Keystore. 
+            // Changeset found for google source to address the related issue with 
+            // remote service accessing keystore :
+            // https://android.googlesource.com/platform/external/sepolicy/+/0e30164b17af20f680635c7c6c522e670ecc3df3
+            // The thrown exception in this case is:
+            // java.lang.NullPointerException: Attempt to invoke interface method 
+            // 'int android.security.IKeystoreService.exist(java.lang.String, int)' on a null object reference
+            // To avoid app from crashing, re-throw as checked exception
+            ClientAnalytics.logEvent(new AndroidKeyStoreFailureEvent(
+                    new InstrumentationPropertiesBuilder(exception)));
+            throw new KeyStoreException(exception);
+        }
+        
+        return isKeyStoreCertAliasExisted;
+    }
 
-        byte[] encrypted = cipher.doFinal(bytes);
+    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
+    private AlgorithmParameterSpec getKeyPairGeneratorSpec(Context context, Date start, Date end) {
+        final String certInfo = String.format(Locale.ROOT, "CN=%s, OU=%s", KEY_STORE_CERT_ALIAS,
+                context.getPackageName());
+        return new KeyPairGeneratorSpec.Builder(context)
+                .setAlias(KEY_STORE_CERT_ALIAS)
+                .setSubject(new X500Principal(certInfo))
+                .setSerialNumber(BigInteger.ONE)
+                .setStartDate(start)
+                .setEndDate(end)
+                .build();
+    }
 
-        // Mac output to sign encryptedData+IV. Keyversion is not included
-        // in the digest. It defines what to use for Mac Key.
-        mac.init(mMacKey);
-        mac.update(blobVersion);
-        mac.update(encrypted);
-        mac.update(iv);
-        byte[] macDigest = mac.doFinal();
+    private SecretKey getSecretKey(byte[] rawBytes) {
+        if (rawBytes != null)
```

Comment:
nit: braces around single line conditional seems to be the convention in this file